### PR TITLE
Nonseekable DataWriter Support

### PR DESCRIPTION
**Pull request by kokorin**
(cherry picked from commit 8d454ec)

### CODE REVIEW

**Comment by mjameson-se** [comment]
Is this change intended to support "live streaming" (http://matroska-org.github.io/matroska-specification/streaming.html)? If so, the Cues and MetaSeek elements should be omitted altogether; as it is with this commit MetaSeek would contain no pointers, rendering the Cues (positioned at the end) nearly useless.

**Comment by kokorin** [comment]
No, this change isn't targeted at livestreaming.
I'm working on command line wrapper for ffmpeg and ffmpeg supports stdin and stdout interaction. So I want to pass programatically created MKV file (uncompressed) via stdin to ffmpeg. Stdin is essentialy unseekable.

**Comment by mjameson-se** [comment]
An interesting use-case... I rather suspect FFMPEG treats this as a "live" stream, since the defining characteristic of a live stream is the lack of seek support, and you might be well advised to observe the suggestions that are made in the spec with respect such streams. 
In particular:
> On the player side, it is possible to detect that a stream is not seekable. If the stream does not have a Meta Seek list or a Cues list at the beginning of the stream, it SHOULD be considered as non seekable

And, due to the inability to update the Segment size after the fact:

> A live Matroska stream is different than a file, because it may have no known end (only when the client disconnects). For that the Segment MUST use the “unknown” size (all 1s in the size).

**Comment by kokorin** [comment]
ffmpeg supports following usage
`cat some.mkv | ffmpeg -f matroska -i - output.mp4`
So it expects normal files in stdin. Also if I try to mimic livestream, shouldn't I also provide frames at live rate? E.g. won't it consume at max 30-60 frames per second?

Anyway, we are discussing appliance of this pull-request in real life, but the problem is that JEBML doesn't respect isSeekable() method.

**Comment by mjameson-se** [comment]
Let me back up a step: my contention is that if the DataSource is not seekable, we should not include MetaSeek or Cues elements because we cannot put enough information in the stream to actually use them effectively. Your example proves that ffmpeg can handle a fully-formed file from stdin, but not that it uses the MetaSeek or Cues elements.

**Review by mjameson-se — CHANGES_REQUESTED** [review]
If the DataSource is not seekable, we cannot correctly write the MetaSeek element which makes Cues effectively unusable. Those elements should be omitted if the DataSource is not seekable.

**Comment by kokorin** [comment]
Done, thanks for clarification. 
BTW I have succeeded in incorporating of JEBML into my [project](/kokorin/Jaffree) to consume and produce frame by frame.

File: src/main/java/org/ebml/matroska/MatroskaFileWriter.java [hunk 1]
```diff
@@ -35,37 +38,84 @@
...
 
   /**
    * @param outputDataWriter DataWriter to write out to.
    */
   public MatroskaFileWriter(final DataWriter outputDataWriter)
   {
     ioDW = outputDataWriter;
+  }
+
+  void initialize() {
```
Comment (mjameson-se) [inline src/main/java/org/ebml/matroska/MatroskaFileWriter.java:62]: For consistency with the style of the existing code, the opening brace should be on its own line. This applies to several other places below.

**Review by mjameson-se — APPROVED** [review]

**Pull request merged**